neqo-transport/src/frame.rs

```diff
@@ -18,68 +18,152 @@ use crate::{
+            0x1 => Ok(Self::Ping),
+            0x2 => Ok(Self::Ack),
+            0x3 => Ok(Self::AckEcn),
+            0x4 => Ok(Self::ResetStream),
+            0x5 => Ok(Self::StopSending),
+            0x6 => Ok(Self::Crypto),
+            0x7 => Ok(Self::NewToken),
+            0x8 => Ok(Self::Stream),
+            0x9 => Ok(Self::StreamWithFin),
+            0xa => Ok(Self::StreamWithLen),
+            0xb => Ok(Self::StreamWithLenFin),
+            0xc => Ok(Self::StreamWithOff),
+            0xd => Ok(Self::StreamWithOffFin),
+            0xe => Ok(Self::StreamWithOffLen),
+            0xf => Ok(Self::StreamWithOffLenFin),
+            0x10 => Ok(Self::MaxData),
+            0x11 => Ok(Self::MaxStreamData),
+            0x12 => Ok(Self::MaxStreamsBiDi),
+            0x13 => Ok(Self::MaxStreamsUniDi),
+            0x14 => Ok(Self::DataBlocked),
+            0x15 => Ok(Self::StreamDataBlocked),
+            0x16 => Ok(Self::StreamsBlockedBiDi),
+            0x17 => Ok(Self::StreamsBlockedUniDi),
+            0x18 => Ok(Self::NewConnectionId),
+            0x19 => Ok(Self::RetireConnectionId),
+            0x1a => Ok(Self::PathChallenge),
+            0x1b => Ok(Self::PathResponse),
+            0x1c => Ok(Self::ConnectionCloseTransport),
+            0x1d => Ok(Self::ConnectionCloseApplication),
+            0x1e => Ok(Self::HandshakeDone),
+            0xaf => Ok(Self::AckFrequency),
+            0x30 => Ok(Self::Datagram),
+            0x31 => Ok(Self::DatagramWithLen),
+            _ => Err(Error::UnknownFrameType),
         }
     }
+}
 
-    const fn from_type_bit(bit: u64, code: u64) -> Self {
-        if (bit & 0x01) == 0 {
-            Self::Transport(code)
-        } else {
-            Self::Application(code)
+impl FrameType {
+    const fn is_stream_with_length(self) -> bool {
+        matches!(
+            self,
+            Self::StreamWithLen
+                | Self::StreamWithLenFin
+                | Self::StreamWithOffLen
+                | Self::StreamWithOffLenFin
+        )
+    }
+
+    const fn is_stream_with_offset(self) -> bool {
+        matches!(
+            self,
+            Self::StreamWithOff
+                | Self::StreamWithOffFin
+                | Self::StreamWithOffLen
+                | Self::StreamWithOffLenFin
+        )
+    }
+    const fn is_stream_with_fin(self) -> bool {
+        matches!(
+            self,
+            Self::StreamWithFin
+                | Self::StreamWithLenFin
+                | Self::StreamWithOffFin
+                | Self::StreamWithOffLenFin
+        )
+    }
+}
+
+impl TryFrom<FrameType> for StreamType {
+    type Error = Error;
+
+    fn try_from(value: FrameType) -> Result<Self, Self::Error> {
+        match value {
+            FrameType::MaxStreamsBiDi | FrameType::StreamsBlockedBiDi => Ok(Self::BiDi),
+            FrameType::MaxStreamsUniDi | FrameType::StreamsBlockedUniDi => Ok(Self::UniDi),
+            _ => Err(Error::FrameEncodingError),
```

Comment:
Is this the right error?  I'm not sure about usage though: what cases lead to this error?

Reply:
We could also use `Error::DecodingFrame`? Not quite sure about the difference in meaning. 

We shouldn't hit this error unless there is a pilot error, i.e., calling trying to convert a `FrameType` into a `StreamType` for which that doesn't make sense (everything other than `MaxStreams` and `StreamsBlocked`).

Reply:
`Error::DecodingFrame` is only used in one place. I would suggest just changing that over to `Error::FrameEncodingError`.

Reply:
Yeah, roll them into one.